include/swift/AST/ArgumentList.h

```diff
@@ -0,0 +1,649 @@
+    const ArgumentList *ArgList;
+    unsigned Idx;
+
+  public:
+    using iterator_category = std::random_access_iterator_tag;
+    using value_type = Argument;
+    using pointer = const value_type *;
+    using reference = value_type;
+    using difference_type = unsigned;
+
+    iterator() : ArgList(nullptr), Idx(0) {}
+    iterator(const ArgumentList *argList, unsigned idx)
+        : ArgList(argList), Idx(idx) {}
+
+    Argument operator*() const {
+      assert(ArgList);
+      return ArgList->get(Idx);
+    }
+    Argument operator[](unsigned i) const {
+      assert(ArgList);
+      return ArgList->get(Idx + i);
+    }
+
+    friend bool operator==(const iterator &lhs, const iterator &rhs) {
+      assert(lhs.ArgList == rhs.ArgList);
+      return lhs.Idx == rhs.Idx;
+    }
+    friend bool operator!=(const iterator &lhs, const iterator &rhs) {
+      return !(lhs == rhs);
+    }
+
+    iterator &operator++() { // pre-increment
+      ++Idx;
+      return *this;
+    }
+    iterator operator++(int) { // post-increment
+      auto tmp = *this;
+      ++*this;
+      return tmp;
+    }
+    iterator &operator--() { // pre-decrement
+      --Idx;
+      return *this;
+    }
+    iterator operator--(int) { // post-decrement
+      auto tmp = *this;
+      --*this;
+      return tmp;
+    }
+
+    iterator &operator+=(unsigned i) {
+      Idx += i;
+      return *this;
+    }
+    iterator operator+(unsigned i) const {
+      return iterator(ArgList, Idx + i);
+    }
+    iterator &operator-=(unsigned i) {
+      Idx -= i;
+      return *this;
+    }
+    iterator operator-(unsigned i) const {
+      return iterator(ArgList, Idx - i);
+    }
+
+    unsigned operator-(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx - rhs.Idx;
+    }
+    bool operator<(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx < rhs.Idx;
+    }
+    bool operator<=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx <= rhs.Idx;
+    }
+    bool operator>(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx > rhs.Idx;
+    }
+    bool operator>=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx >= rhs.Idx;
+    }
+  };
+
+  iterator begin() const { return {this, 0}; }
+  iterator end() const { return {this, size()}; }
+
+  Argument front() const { return get(0); }
+  Argument back() const { return get(size() - 1); }
+
+  Argument operator[](unsigned idx) const { return get(idx); }
+
+  /// Returns the index of the first trailing closure in the argument list, or
+  /// \c None if there are no trailing closures.
+  ///
+  /// Note that in type-checked AST, this may return an index of a variadic
+  /// expansion which contains the trailing closure (and it may not necessarily
+  /// be the first argument of the expansion).
+  Optional<unsigned> getRawFirstTrailingClosureIndex() const {
```

Comment:
I was under the impression that the word "raw" referred to the fact that we were storing it in this weird format where `NumArgs` means `None`, but here you're calling it `getRawFirstTrailingClosureIndex()` even though you're converting it from that format. So what does "raw" mean in this name?

Reply:
The intention behind using "raw" was to communicate that the index you get might not be quite what you're expecting for type-checked AST (e.g you shouldn't in general do `auto numTrailing = argList->size() - *argList->getRawFirstTrailingClosureIndex()`). I struggled to think of another name that would communicate this, do you have any suggestions?

Alternatively we could rename the stored property to something like `FirstTrailingClosureIndexBits` to avoid conflating the meaning of "raw" here.

Reply:
Hmm. I'm trying to understand this…

* The comment mentions varargs; I think it's saying that the first trailing closure argument might be in a `VarargExpansionExpr`. Is that correct? If so, what are you supposed to do in this situation? Is the `VarargExpansionExpr` going to somehow communicate to you which element is the first trailing closure argument?

* Are there other situations where `getRawFirstTrailingClosureIndex()` is "wrong"?

Reply:
> The comment mentions varargs; I think it's saying that the first trailing closure argument might be in a `VarargExpansionExpr`. Is that correct?

Yup, that's correct

> If so, what are you supposed to do in this situation? Is the `VarargExpansionExpr` going to somehow communicate to you which element is the first trailing closure argument?

Yeah, unless you look at source ranges, you wouldn't be able to know. It could be that there's some valid use case from looking at the first argument that contains a trailing closure, but it seems unlikely. As said in my other comment, I'm starting to think the right answer here might be to completely forbid calling this on a type-checked argument list.

> Are there other situations where `getRawFirstTrailingClosureIndex()` is "wrong"?

I don't believe so, it should always point to the first argument that is trailing outside of the r-paren source range. Though there are still other footguns with it, e.g default arguments and regular arguments that have been re-ordered may appear after it.

Reply:
It's probably worth asking, what's the purpose of recording which arguments are "trailing closure arguments"? If it's so that the typechecker knows which arguments should be typechecked specially (e.g. loose argument label matching), you could argue that the typechecker basically turns trailing closure arguments into normal arguments. Thus, any (successfully) typechecked argument list will inherently have no trailing closure arguments.

Reply:
As far as I can tell the trailing closure info serves two main purposes:

1. For the expression type-checker, including solving and diagnostics
2. For various IDE related logic

The former mostly operates on un-type-checked AST (with the exception of MiscDiagnostics), so should be able to query trailing closure info from the argument list with no problems. The latter mostly operates on type-checked AST, so should always go through `getOriginalArguments()`.

> you could argue that the typechecker basically turns trailing closure arguments into normal arguments. Thus, any (successfully) typechecked argument list will inherently have no trailing closure arguments.

I really like the cleanness of that model, though I do worry that code might accidentally query trailing closure info on a type-checked argument list and not get the answer it's expecting (as it probably should have queried `getOriginalArguments()` instead). So I feel like some kind of warning, or assertion failure if you call it on type-checked AST would be useful there.

Reply:
Updated to force users to query through `getOriginalArgs()` for type-checked argument lists, as I feel it's important to avoid accidentally querying the wrong trailing closure info, and it's what the IDE + MiscDiagnostic logic tends to want anyways, how does this look?

Reply:
I like it!